lahja/endpoint.py

```diff
@@ -90,6 +95,43 @@ class ProxyEndpointConnector(BaseProxy):  # type: ignore # TypeShed is missing B
     def put_nowait(self, item_and_config: Tuple[BaseEvent, Optional[BroadcastConfig]]) -> None:
         self._callmethod('put_nowait', (item_and_config,))
 
+    def subscribed_events(self) -> Set[Type[BaseEvent]]:
+        return self._callmethod('subscribed_events')
+
+
+class ConnectorFilter:
+    """
+    Filters out messages the remote endpoint doesn't want to receive
+    """
+
+    def __init__(self, proxy: ProxyEndpointConnector, condition: ConditionProxy, loop):
+        self.proxy = proxy
+        self.condition = condition
+        self.loop = loop
+
+        self.update_subscription()
+
+    def start(self):
+        # TODO: we probably also want stop()
+        threading.Thread(target=self._run, daemon=True).start()
+
+    def _run(self):
+        while True:
+            with self.condition:
+                self.update_subscription()
+                self.condition.wait()
+            self.update_subscription()
+
+    def update_subscription(self):
+        self.allowed_events = self.proxy.subscribed_events()
+
+    def put_nowait(self, item_and_config: Tuple[BaseEvent, Optional[BroadcastConfig]]) -> None:
+        item, config = item_and_config
+        is_response = config is not None and config.filter_event_id
+        is_request = hasattr(item, '_id')
```

Comment:
Yeah, I agree that using `_id` isn't ideal! In my defense, `BaseEvent` also assumes it :) Some of the tests do too, so if this were to change it would be caught pretty quickly.

https://github.com/ethereum/lahja/blob/e7928a8935fc77be332028b970fc6ae261970f87/lahja/misc.py#L42-L58

It looks like `filter_endpoint` doesn't work though, nothing enforces that it will be set. For example, it's not set in any of lahja's tests! I think I have a solution though, updates incoming.

Reply:
It's used in this test.

https://github.com/ethereum/lahja/blob/db081141b7184429ff60bd9e1673a1483773c2ee/tests/core/test_broadcast_config.py#L65

Reply:
> It's used in this test.

Heh, oops, s/any/many/, good catch!

> Ok, curious what you come up with!

I spent some time thinking about this yesterday, and I can't find any obviously right solution.

`is_request` really shouldn't need to be there. If there's nothing on the other side listening for this request then there's no point in sending the request! A future PR could even notice that `filter_endpoint` was used and throw an exception rather than block the calling coro forever while it waits for a response which will never come.

However, removing `is_request` causes the tests to fail, and this is because of an annoying race condition: 

https://github.com/ethereum/lahja/blob/3a9a107ea6d7be893b0e2526838b95aed5f384d4/tests/core/test_basics.py#L17-L28

Here `endpoint.request()` blocks forever because this thread moves too quickly. There's no time for `notify_all()` to tell the endpoint about the new subscription or for the endpoint to react. This can be fixed by adding an `await asyncio.sleep(0.05)`, but that'd have to be added to many of the tests.

I don't think this will be a problem in the real world. I don't think we have anything which subscribes and expects to immediately receive messages, where a dropped initial message would cause problems. Since everything is in different processes trinity doesn't have the same idea of simultaneity that these lahja tests have.

There are some options for dealing with this:
1. Just add `asyncio.sleep()` to the tests. There's good precedent here, we often call `asyncio.sleep(0)` to allow other coroutines a chance to run.
2. Add a helper, `await_subscriptions_updated(endpoint, event_type)`, which blocks until the endpoint notices the subscription.
3. Somehow make `_notify_subscriptions_changed` synchronous. I couldn't find an easy way to do this, there's not really a way for `Server`s to send messages back to their clients.
4. Somehow make an `asyncio` version of `BaseManager` such that while everything is in the same process we don't need to use any extra threads, and `asyncio.sleep(0)` is enough to register a subscription.
5. Give our Endpoints the ability to inspect the `BaseManager` and notice that the remote Endpoint is running in the same thread as we are. When that's true `notify_subscriptions_changed` does something different which is synchronous, so we never have a race condition.

I don't like any of these, do you have any thoughts?

Reply:
>If there's nothing on the other side listening for this request then there's no point in sending the request!

Correct!

>I don't think this will be a problem in the real world. I don't think we have anything which subscribes and expects to immediately receive messages, where a dropped initial message would cause problems. Since everything is in different processes trinity doesn't have the same idea of simultaneity that these lahja tests have.

Yes, I agree, there shouldn't be a guarantee about which message will be the initial one receives after subscribing.

Solution 2. would be my favorite as well but I'm easy to please so I would even be ok with 1. :smile: